Add (optional) metrics reporting

lstoll opened the pull request:
Reports some basic data, like error/success and data age.

Serve updated to use a rungroup to manage the various processes, better
shutdown handling added via that.

lstoll committed:
Add (optional) metrics reporting

Reports some basic data, like error/success and data age.

Serve updated to use a rungroup to manage the various processes, better
shutdown handling added via that.

sr commented on main.go:
```diff
@@ -263,31 +278,111 @@ func main() {
+			})
+
+			metrDone := make(chan struct{}, 1)
+			g.Add(func() error {
```
This could be implemented as a prometheus.Collector (e.g. https://github.com/prometheus/client_golang/blob/master/prometheus/go_collector.go)

lstoll commented:
So I guess in that case, it’d just refresh metrics on scrape? Seems better, I’ll give it a shot.

> On Jul 27, 2020, at 08:49, Simon Rozet <notifications@github.com> wrote:
> 
> ﻿
> @sr commented on this pull request.
> 
> In main.go:
> 
> > +				if err := metricsSrv.ListenAndServe(); err != nil {
> +					return fmt.Errorf("serving metrics: %v", err)
> +				}
> +				return nil
> +			}, func(error) {
> +				ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
> +				defer cancel()
> +				if err := metricsSrv.Shutdown(ctx); err != nil {
> +					log.Printf("shutting down metrics server: %v", err)
> +				}
> +				log.Print("returning metrics shutdown")
> +
> +			})
> +
> +			metrDone := make(chan struct{}, 1)
> +			g.Add(func() error {
> This could be implemented as a prometheus.Collector (e.g. https://github.com/prometheus/client_golang/blob/master/prometheus/go_collector.go)
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

sr commented:
> So I guess in that case, it’d just refresh metrics on scrape? Seems better, I’ll give it a shot.

Exactly. That's the recommended approach:

> Metrics should only be pulled from the application when Prometheus scrapes them, exporters should not perform scrapes based on their own timers. That is, all scrapes should be synchronous.

https://prometheus.io/docs/instrumenting/writing_exporters/#scheduling

lstoll committed:
move to prometheus.Collector for on-demand metrics

Rather than a background task to update things, implement a collector to fetch
data on demand. this is the "correct" way.

lstoll commented:
@sr moved to a `prometheus.Collector`, wdyt?

lstoll committed:
whitespace

sr commented on metrics.go:
```diff
@@ -0,0 +1,88 @@
+
+	lt, err := m.s.LatestLocationTimestamp(ctx)
+	if err != nil {
+		m.l.Printf("metrics: failed to get latest device location timestamp: %v", err)
```
Could use `prometheus.NewInvalidMetric` here, if you pass in a logger to `promhttp.HandlerFor` it'll log those errors:

https://github.com/sr/repo/blob/master/src/aux1/cmd/laptop-manager/swinsian.go#L93
https://github.com/prometheus/client_golang/blob/master/prometheus/promhttp/http.go#L129

sr reviewed (approved):
Nice

lstoll committed:
report metrics errors properly

lstoll commented:
Rad, that’s better. Updated with that. Why read the docs when I can just wait for feedback

> On Jul 29, 2020, at 09:15, Simon Rozet <notifications@github.com> wrote:
> 
> 
> @sr approved this pull request.
> 
> Nice
> 
> In metrics.go <https://github.com/lstoll/wherewasi/pull/17#discussion_r462332393>:
> 
> > +}
> +
> +func (m *metricsCollector) Describe(ch chan<- *prometheus.Desc) {
> +	ch <- m.lastDeviceLocationTime
> +	ch <- m.latestFoursquareCheckin
> +}
> +
> +func (m *metricsCollector) Collect(ch chan<- prometheus.Metric) {
> +	// we don't get a context from the scrape here, so just create one with a
> +	// reasonably timeout to avoid blocking things here
> +	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
> +	defer cancel()
> +
> +	lt, err := m.s.LatestLocationTimestamp(ctx)
> +	if err != nil {
> +		m.l.Printf("metrics: failed to get latest device location timestamp: %v", err)
> Could use prometheus.NewInvalidMetric here, if you pass in a logger to promhttp.HandlerFor it'll log those errors:
> 
> https://github.com/sr/repo/blob/master/src/aux1/cmd/laptop-manager/swinsian.go#L93 <https://github.com/sr/repo/blob/master/src/aux1/cmd/laptop-manager/swinsian.go#L93>
> https://github.com/prometheus/client_golang/blob/master/prometheus/promhttp/http.go#L129 <https://github.com/prometheus/client_golang/blob/master/prometheus/promhttp/http.go#L129>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/lstoll/wherewasi/pull/17#pullrequestreview-457565349>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAAAFNVUMOFSNWGJ4EH7QADR6AVIVANCNFSM4PIFMJOQ>.
>
